### PR TITLE
Feat/midas predictor en

### PR DESCRIPTION
**Pull request by dilyararimovna**

### CODE REVIEW

File: annotators/midas_predictor/model_training.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import tensorflow_hub as hub
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.metrics import f1_score, accuracy_score
+from sklearn.model_selection import train_test_split
+
+from data2seq import Dial2seq, SequencePreprocessor
+from midas_dataset import MidasDataset, MidasVectorizer
+
+
+# topical_sequencer = Dial2seq('data/topical_chat_annotated.json', 3)
```
Comment (deniskapel) [inline annotators/midas_predictor/model_training.py:15]: uncomment

File: annotators/midas_predictor/model_training.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+if USE_MODEL_PATH is None:
+    USE_MODEL_PATH = "https://tfhub.dev/google/universal-sentence-encoder/4"
+
+encoder = hub.load(USE_MODEL_PATH)
+
+midas_vectorizer = MidasVectorizer(
+    text_vectorizer=encoder, midas2id=Midas2Id, context_len=3, embed_dim=512  # USE  # USE vector size
+)
+
+midas_vectorizer.context_vector(train[0]["previous_text"], train[0]["midas_vectors"])
```
Comment (deniskapel) [inline annotators/midas_predictor/model_training.py:100]: can be removed, just tests that vectorizer works properly

File: annotators/midas_predictor/model_training.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+test_dataloader = MidasDataset(data=test, vectorizer=midas_vectorizer, batch_size=len(test), shuffle=False)
+
+for X_train, y_train in train_dataloader:
+    break
+
+for X_val, y_val in val_dataloader:
+    break
+
+for X_test, y_test in test_dataloader:
```
Comment (deniskapel) [inline annotators/midas_predictor/model_training.py:114]: test split can be removed from here. We have it for toloka only.

File: annotators/midas_predictor/midas_dataset.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
+import numpy as np
+from tensorflow.keras.utils import Sequence
+
+
+class MidasDataset(Sequence):
+    """customized Dataset class from torch"""
```
Comment (deniskapel) [inline annotators/midas_predictor/midas_dataset.py:6]: from tensorflow

File: annotators/midas_predictor/data2seq.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
+import json
+from collections import Counter
+
+
+class Dial2seq:
+    """
+    a class to transform dialogues into a sequence of utterances and labels
+    The sequence consists of n previous uterrances.
+
+    There are no constraints
```
Comment (deniskapel) [inline annotators/midas_predictor/data2seq.py:10]: This paragraph can be removed

File: annotators/midas_predictor/model_training.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    text_vectorizer=encoder, midas2id=Midas2Id, context_len=3, embed_dim=512  # USE  # USE vector size
+)
+
+midas_vectorizer.context_vector(train[0]["previous_text"], train[0]["midas_vectors"])
+
+train_dataloader = MidasDataset(data=train, vectorizer=midas_vectorizer, batch_size=len(train), shuffle=False)
+
+val_dataloader = MidasDataset(data=val, vectorizer=midas_vectorizer, batch_size=len(val), shuffle=False)
+
+test_dataloader = MidasDataset(data=test, vectorizer=midas_vectorizer, batch_size=len(test), shuffle=False)
```
Comment (deniskapel) [inline annotators/midas_predictor/model_training.py:106]: test split can be removed from here. We have it for toloka only.

File: annotators/midas_predictor/data2seq.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        and return a midas vector per each sentence
+        """
+        labels = []
+        vectors = []
+
+        for sample in midas_labels:
+            labels.append(max(sample[0], key=sample[0].get))
+            vectors.append(list(sample[0].values()))
+
+        return labels, vectors
```
Comment (deniskapel) [inline annotators/midas_predictor/data2seq.py:171]:
```suggestion
class SequencePreprocessor:
    """ 
    Prepares a dataset for training - can be updated with filters,
    e.g., to filter out invalid sequences
    """
    def transform(self, sequences: list) -> list:
        """ extract necessary data from sequences """
        seqs = list()
        
        for seq in sequences:
            sample = self.__get_dict_entry(self.__shape_output(seq))
            seqs.append(sample)
            
        return seqs
    
    def __shape_output(self, seq: list) -> list:
        """ shapes sequence in order to keep only the necessary data """
        
        output = list()
        
        for ut in seq[:-1]:
                
            midas_labels, midas_vectors = self.__get_midas(ut['midas'])
            
            output.append((ut['text'], midas_labels, midas_vectors))

        # preprocess only the first sentence of 
        # the last utterance in the sequence
        midas_labels, midas_vectors = self.__get_midas(seq[-1]['midas'])
        midas_labels, midas_vectors = midas_labels[0:1], midas_vectors[0:1]
        sentence = seq[-1]['text'][0]
            
        output.append(
            (sentence, midas_labels[0:1]))
        
        return output
    
    def __get_dict_entry(self, seq) -> dict:
        """ creates a proper dict entry to dump into a file """
        entry = dict()
        entry['previous_text'] = [s[0] for s in seq[:-1]]
        entry['previous_midas'] = [s[1] for s in seq[:-1]]
        entry['midas_vectors'] = [s[2] for s in seq[:-1]]
        entry['predict'] = {}
        entry['predict']['text'] = seq[-1][0]
        entry['predict']['midas'] = seq[-1][1][0]
        
        return entry
            
        
    def __get_midas(self, midas_labels: list) -> tuple:
        """ 
        extracts midas labels with max value per each sentence in an utterance
        and return a midas vector per each sentence
        """
        labels = []
        vectors = []
        
        for sample in midas_labels:
            labels.append(max(sample[0], key=sample[0].get))
            vectors.append(list(sample[0].values()))
            
        return labels, vectors
```

Comment (deniskapel) [inline annotators/midas_predictor/data2seq.py:171]:
```suggestion
class SequencePreprocessor:
    """ 
    Prepares a dataset for training - can be updated with filters,
    e.g., to filter out invalid sequences
    """
    def transform(self, sequences: list) -> list:
        """ extract necessary data from sequences """
        seqs = list()
        
        for seq in sequences:
            sample = self.__get_dict_entry(self.__shape_output(seq))
            seqs.append(sample)
            
        return seqs
    
    def __shape_output(self, seq: list) -> list:
        """ shapes sequence in order to keep only the necessary data """
        
        output = list()
        
        for ut in seq[:-1]:
                
            midas_labels, midas_vectors = self.__get_midas(ut['midas'])
            
            output.append((ut['text'], midas_labels, midas_vectors))

        # preprocess only the first sentence of 
        # the last utterance in the sequence
        midas_labels, midas_vectors = self.__get_midas(seq[-1]['midas'])
        midas_labels, midas_vectors = midas_labels[0:1], midas_vectors[0:1]
        sentence = seq[-1]['text'][0]
            
        output.append(
            (sentence, midas_labels[0:1]))
        
        return output
    
    def __get_dict_entry(self, seq) -> dict:
        """ creates a proper dict entry to dump into a file """
        entry = dict()
        entry['previous_text'] = [s[0] for s in seq[:-1]]
        entry['previous_midas'] = [s[1] for s in seq[:-1]]
        entry['midas_vectors'] = [s[2] for s in seq[:-1]]
        entry['predict'] = {}
        entry['predict']['text'] = seq[-1][0]
        entry['predict']['midas'] = seq[-1][1][0]
        
        return entry
            
        
    def __get_midas(self, midas_labels: list) -> tuple:
        """ 
        extracts midas labels with max value per each sentence in an utterance
        and return a midas vector per each sentence
        """
        labels = []
        vectors = []
        
        for sample in midas_labels:
            labels.append(max(sample[0], key=sample[0].get))
            vectors.append(list(sample[0].values()))
            
        return labels, vectors
```

File: annotators/midas_predictor/data2seq.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+class Dial2seq:
+    """
+    a class to transform dialogues into a sequence of utterances and labels
+    The sequence consists of n previous uterrances.
+
+    There are no constraints
+    on number of entities or midas types in the them, however a sequence is
+    valid only when it is followed by an utterance with a single entity and
+    its midas label is not MISC or ANAPHOR.
```
Comment (deniskapel) [inline annotators/midas_predictor/data2seq.py:13]: can be removed

File: annotators/midas_predictor/data2seq.py (hunk 7 above)
Comment (deniskapel) [reply annotators/midas_predictor/data2seq.py:171]: entities are not valid for this predictor, can be removed

File: annotators/midas_predictor/data2seq.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+    params:
+    path2data: str - a path to a json file with dialogues and their midas and cobot labels
+    seqlen: int - a number of utterances to use to predict next midas labels and entities
+    """
```
Comment (deniskapel) [inline annotators/midas_predictor/data2seq.py:18]:
```suggestion
    """
    a class to transform dialogues into a sequence of utterances and labels
    The sequence consists of n previous uterrances.

    params:
    path2data: str - a path to a json file with dialogues and their midas annotation
    seqlen: int - a number of utterances to use to predict next midas labels, i.e. context length
    """
```

**Review by deniskapel — COMMENTED** [review]
suggest several changes to training files to data2seq.py midas_dataset.py and model_training.py to simplify and not mix with entities

File: annotators/midas_predictor/model_training.py (hunk 3 above)
Comment (dilyararimovna) [reply annotators/midas_predictor/model_training.py:114]: left only to see data sizes

**Review by deniskapel — APPROVED** [review]

File: annotators/midas_predictor/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+RUN pip install -r /src/requirements.txt
+
+COPY ./ /src/
+COPY ./common/ /src/common/
+WORKDIR /src
+
+RUN mkdir /data/
+RUN mkdir /data/models/
+
+RUN wget -c -q http://lnsigo.mipt.ru/export/dream/midas_predictor_rfc_depth20.pickle -P /data/models/
```
Comment (IgnatovFedor) [inline annotators/midas_predictor/Dockerfile:18]: Почему не files.deeppavlov.ai?

**Pull request merged**